mutation/index.ts

```diff
@@ -0,0 +1,132 @@
+import { useCallback, useRef } from 'react'
+import useSWR, {
+  useSWRConfig,
+  Middleware,
+  Key,
+  Fetcher,
+  SWRResponse
+} from 'swr'
+
+import { serialize } from '../src/utils/serialize'
+import { useStateWithDeps } from '../src/utils/state'
+import { withMiddleware } from '../src/utils/with-middleware'
+import { useIsomorphicLayoutEffect } from '../src/utils/env'
+import { isUndefined, UNDEFINED } from '../src/utils/helper'
+import { getTimestamp } from '../src/utils/timestamp'
+
+import { SWRMutationConfiguration, SWRMutationResponse } from './types'
+
+const mutation = <Data, Error>() => (
+  key: Key,
+  fetcher: Fetcher<Data>,
+  config: SWRMutationConfiguration<Data, Error> = {}
+) => {
+  const { mutate } = useSWRConfig()
+
+  const keyRef = useRef(key)
+  // Ditch all mutation results that happened earlier than this timestamp.
+  const ditchMutationsTilRef = useRef(0)
+
+  const [stateRef, stateDependencies, setState] = useStateWithDeps<Data, Error>(
+    {
+      data: UNDEFINED,
+      error: UNDEFINED,
+      isValidating: false
+    },
+    true
+  )
+  const currentState = stateRef.current
+
+  // Similar to the global mutate, but bound to the current cache and key.
+  // `cache` isn't allowed to change during the lifecycle.
+  const boundMutate: SWRResponse<Data, Error>['mutate'] = useCallback(
+    (newData, shouldRevalidate) => {
+      const serializedKey = serialize(keyRef.current)[0]
+      return mutate(serializedKey, newData, shouldRevalidate)
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    []
+  )
+
+  const trigger = useCallback(async (extraArg, opts) => {
+    if (!fetcher) {
+      throw new Error('Can’t trigger the mutation: missing fetcher.')
```

Comment:
Do you prefer type-hint rather than error message here? If it's not defined it will crash anyway and we can hint user by `Required<Type>` type in the function declaration

Reply:
Yeah that would be better!